### PR TITLE
Use a settings object for the main settings

### PR DESCRIPTION
**Pull request by ShadowNinja**
Usage:

``` Lua
minetest.settings:get("foo")
```

This unifies the settings APIs and adds `LuaSettings::l_set_bool()` (and unifies the API module initialization APIs (`Initialize()`/`InitializeAsync()`)).

### CODE REVIEW

**Comment by ShadowNinja** [comment]
@kahrl I fixed that problem.

**Comment by ShadowNinja** [comment]
Build failed due to a bad key server.
That's happened a lot recently, the key server should probably be changed.

**Comment by ShadowNinja** [comment]
Rebased and updated to support security.

**Comment by nerzhul** [comment]
Jenkins build passed

**Comment by paramat** [comment]
Needs reviews.

**Comment by ShadowNinja** [comment]
Note: even if the settings change isn't merged, the ModApi::Initialize[Async] unification is worth looking into on its own.

**Comment by celeron55** [comment]
I would guess logging a warning for every setting_get/setting_getbool is going to cause too much logging in some cases.

Otherwise this seems fine to me.

**Comment by ShadowNinja** [comment]
Rebased, and added docs.
I also fixed some things in the profiler that were obviously wrong (like `if minetest.setting_get("foo") or true then`).

**Comment by nerzhul** [comment]
Sounds interesting but please remove many unneeded noise

**Comment by celeron55** [comment]
There's some weird builtin/game/mod_profiling.lua file in this now?

**Comment by ShadowNinja** [comment]
@celeron55: Oops, rebase error.

**Comment by celeron55** [comment]
Looks fine now. Haven't tested it though.

**Comment by rubenwardy** [comment]
I'd prefer it to match the meta data api. Ie: get_string rather than get.

**Comment by ShadowNinja** [comment]
@rubenwardy: That would mean changing the existing Settings API, which would be a separate issue.

**Comment by rubenwardy** [comment]
You're already changing the Settings API. It would be as simple as renaming get to get_string, and checking the others match

**Comment by ShadowNinja** [comment]
@rubenwardy: No, I'm not changing the API, just having the engine create a new instance of Settings.  And I can't just change it, I'd need compatibility functions.  Finally, I don't see any good reason to change it when `get_string` is longer and isn't much (if any) better.  That change can be done as a separate PR if you want.

File: builtin/game/chatcommands.lua [hunk 1]
```diff
@@ -718,7 +718,7 @@ core.register_chatcommand("rollback", {
 	description = "Revert actions of a player. Default for <seconds> is 60",
 	privs = {rollback=true},
 	func = function(name, param)
-		if not core.setting_getbool("enable_rollback_recording") then
+		if not core.settings:get_bool("enable_rollback_recording") then
```
Comment (rubenwardy) [inline builtin/game/chatcommands.lua]: you renamed getbool to get_bool

Consistency is more important than a long name

Comment (ShadowNinja) [inline builtin/game/chatcommands.lua]: No, I removed `settings_getbool` and replaced it with the existing Settings's `get_bool`.  This is just a preexisting inconsistency between the names of two different functions that I switched between.

Comment (rubenwardy) [inline builtin/game/chatcommands.lua]: ah, I see. That's fine then

File: src/script/lua_api/l_settings.h [hunk 2]
```diff
@@ -54,19 +57,23 @@ class LuaSettings : public ModApiBase
...
 
+	static void create(lua_State *L, Settings *settings, const std::string &filename);
+
 	// LuaSettings(filename)
-	// Creates an LuaSettings and leaves it on top of stack
+	// Creates a LuaSettings and leaves it on top of the stack
 	static int create_object(lua_State *L);
 
-	static LuaSettings *checkobject(lua_State *L, int narg);
+	static LuaSettings* checkobject(lua_State *L, int narg);
```
Comment (rubenwardy) [inline src/script/lua_api/l_settings.h]: ^

**Pull request merged**